src/map/guild.c

```diff
@@ -147,26 +148,70 @@ static int guild_check_skill_require(struct guild *g, int id)
 	return 1;
 }
 
-static bool guild_read_castledb(char *str[], int columns, int current)
-{// <castle id>,<map name>,<castle name>,<castle event>[,<reserved/unused switch flag>]
-	struct guild_castle *gc;
-	int index;
+static bool guild_read_castledb_libconfig(void)
+{
+	struct config_t castle_conf;
+	struct config_setting_t *castle_db = NULL, *it = NULL;
+	const char *config_filename = "db/castle_db.conf"; // FIXME hardcoded name
+	int i = 0;
+
+	if (libconfig->load_file(&castle_conf, config_filename) == 0)
+		return false;
 
-	nullpo_retr(false, str);
-	index = mapindex->name2id(str[1]);
-	if (map->mapindex2mapid(index) < 0) // Map not found or on another map-server
+	if ((castle_db = libconfig->setting_get_member(castle_conf.root, "castle_db")) == NULL) {
+		ShowError("guild_read_castledb_libconfig: can't read %s\n", config_filename);
 		return false;
+	}
+
+	while ((it = libconfig->setting_get_elem(castle_db, i++)) != NULL) {
+		guild->read_castledb_libconfig_sub(it, i - 1, config_filename);
+	}
 
+	libconfig->destroy(&castle_conf);
+	ShowStatus("Done reading '"CL_WHITE"%d"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", i, config_filename);
+	return true;
+}
+
+static bool guild_read_castledb_libconfig_sub(struct config_setting_t *it, int idx, const char *source)
+{
+	nullpo_ret(it);
+	nullpo_ret(source);
+
+	struct guild_castle *gc = NULL;
+	const char *name = NULL;
+	int i32 = 0;
 	CREATE(gc, struct guild_castle, 1);
-	gc->castle_id = atoi(str[0]);
+
+	if (libconfig->setting_lookup_int(it, "CastleID", &i32) == CONFIG_FALSE) {
```

Comment:
This is leaking `gc` (same goes for all the other early returns in this function)